packages/web/src/components/Form.js

```diff
@@ -0,0 +1,191 @@
+// Module imports
+import {
+	createContext,
+	useCallback,
+	useContext,
+	useEffect,
+	useReducer,
+} from 'react'
+import PropTypes from 'prop-types'
+
+
+
+
+
+// Local imports
+import {
+	auth,
+	firestore,
+} from 'helpers/firebase'
+
+
+
+
+
+// Constants
+const INITIAL_STATE = {
+	errors: {},
+	isValid: false,
+	isTouched: false,
+	touched: {},
+	validity: {},
+	values: {},
+}
+
+
+
+
+
+function reducer(state, action) {
+	const {
+		payload,
+		type,
+	} = action
+	const newState = {
+		...INITIAL_STATE,
+		...state,
+	}
+
+	switch (type) {
+		case 'validity changed':
+			newState.validity = {
+				...newState.validity,
+				[payload.fieldName]: !payload.errors?.length,
+			}
+			newState.errors = {
+				...newState.errors,
+				[payload.fieldName]: payload.errors,
+			}
+			newState.isValid = !Object
+				.values(newState.validity)
+				.some(isValid => !isValid)
+			break
+
+		case 'value changed':
+			newState.values = {
+				...newState.values,
+				[payload.fieldName]: payload.value,
+			}
+			newState.touched = {
+				...newState.touched,
+				[payload.fieldName]: newState.initialValues[payload.fieldName] !== payload.value,
+			}
+			newState.isTouched = Object
+				.values(newState.touched)
+				.some(isTouched => isTouched)
+			break
+
+		default:
+			console.warn(`Unrecognized action dispatched: ${type}`, payload)
```

Comment:
Rather than log warnings to the console where they may never be seen, it might be a nice enhancement to log these errors to an error log in firestore

Reply:
Agreed. The problem with logging errors to Firestore, though, is that we're charged per document per read/write. That's not a big deal when writing a few small documents, but when trying to read a huge list of logs, it adds up. I'm sure we could find something like LogRocket or Loggly that offers a free open source license, though.